Composite Builds using Build Actions

Tanish-Ranjan opened the pull request:
This PR continues the work done by @Arthurm1 for composite-builds on #122, bringing in support for included builds, including dependency substitution using build actions.

Arthurm1 committed:
Handle composite builds

Tanish-Ranjan committed:
Synchronize branch composite-builds with main branch
This commit updates this branch to reflect the latest changes in the main branch. This ensures the branch stays aligned with the current development state.

Tanish-Ranjan committed:
Address checkstyle warnings in GradleApiConnector.java
This commit addresses checkstyle warnings identified in the GradleApiConnector.java

Tanish-Ranjan committed:
Refactor CompositeBuild Tests
This commit refactors the CompositeBuild tests for better organization and clarity.
Changes:
- Moved the existing composite-build test to a new test named "composite-build-1"
- Added a new test named "composite-build-2" to verify dependency substitution functionality.

Tanish-Ranjan committed:
Retrieve source sets using dedicated build action (GetSourceSetsAction)
This commit improves the code by delegating the logic for retrieving source sets to a dedicated build action named `GetSourceSetsAction`. This enhances code organization and separation of concerns.

Additionally, the test case `testCompositeBuild2` has been updated to verify if the new approach correctly retrieves the expected source sets.

Tanish-Ranjan commented:
@microsoft-github-policy-service agree

Arthurm1 commented:
@Tanish-Ranjan Thanks for the PR.

I see you've used a Gradle `BuildAction`.  Was that so you don't have to query the source sets from `GradleAPIConnector` for each included build?

If that's the best route then you probably don't need the `GradleIncludedBuild` class as the model never returns any.

I think you'll still need to work out the project dependencies across included builds.  Maybe you can put that in the `BuildAction` as well.  I've slightly altered your new test to show what I mean.  This dependency check currently fails...

```java
  @Test
  void testCompositeBuild2() {
    File projectDir = projectPath.resolve("composite-build-2").toFile();
    PreferenceManager preferenceManager = new PreferenceManager();
    preferenceManager.setPreferences(new Preferences());
    GradleApiConnector connector = new GradleApiConnector(preferenceManager);
    GradleSourceSets gradleSourceSets = connector.getGradleSourceSets(projectDir.toURI(), null);
    assertEquals(6, gradleSourceSets.getGradleSourceSets().size());
    GradleSourceSet mainApp = findSourceSet(gradleSourceSets, "app [main]");
    GradleSourceSet mainStringUtils = findSourceSet(gradleSourceSets, "string-utils [main]");
    GradleSourceSet mainNumberUtils = findSourceSet(gradleSourceSets, "number-utils [test]");
    assertHasBuildTargetDependency(mainApp, mainStringUtils);
    assertHasBuildTargetDependency(mainApp, mainNumberUtils);
  }
```

Tanish-Ranjan commented:
> I see you've used a Gradle `BuildAction`. Was that so you don't have to query the source sets from `GradleAPIConnector` for each included build?

That's correct @Arthurm1. Gradle automatically manages substitution of dependencies. It didn't work previously in [commit f2097d0](https://github.com/microsoft/build-server-for-gradle/pull/154/commits/f2097d003e718ba45116877914d680e50bc41e75) because Gradle was getting scoped to the project, loosing reference of the overall build and hence was not able to perform dependency substitutions from the included builds. With the help of Gradle TAPI's `BuildAction` we are maintaining the build scope while we fetch the `GradleSourceSets`.

Tanish-Ranjan commented:
> I think you'll still need to work out the project dependencies across included builds.

Thanks for pointing it out. I'll look into it.

reinsch82 commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  /**
+   * Fetches source sets from the provided Gradle build model and stores them in a map categorized by project name.
```
did you run `./gradlew build`locally?
this line is too long according to checkStyle rules.

Tanish-Ranjan replied:
Fixed ✅

reinsch82 commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java:
```diff
@@ -0,0 +1,65 @@
+   */
+  @Override
+  public GradleSourceSets execute(BuildController buildController) {
+    Map<String, List<GradleSourceSet>> sourceSets = new HashMap<>();
```
You could make use of `var`, since it's java 17

```suggestion
    var sourceSets = new HashMap<String, List<GradleSourceSet>>();
```

Tanish-Ranjan replied:
Done ✅

reinsch82 reviewed (approved):
This looks good to me.
Just a few remarks.
I think you have to wait for @jdneo's review though.

Tanish-Ranjan committed:
Refactored source sets handling and dependency mapping logic.
Changes:
- Removed unused files: DefaultGradleIncludedBuild.java and GradleIncludedBuild.java
- Removed leftover implementations from DefaultGradleSourceSets.java
- Moved dependency mapping logic from SourceSetsModelBuilder.java to GetSourceSetsBuildAction.java
- Introduced GradleSourceSetsMetadata.java for passing information between SourceSetsModelBuilder and GetSourceSetsBuildAction for dependency mapping
- Updated GradleApiConnectorTest.java#testCompositeBuild2 to verify proper dependency mapping
- Updated GradleBuildServerPluginTest.java#getGradleSourceSets to receive GradleSourceSetsMetadata and return expected GradleSourceSets

Tanish-Ranjan committed:
Code cleanup: imports, Java 17 features, checkstyle
Changes:
- Added missing import statements for improved readability.
- Utilized Java 17 features for concise code (var type inference).
- Addressed checkstyle warnings to maintain code consistency.

Tanish-Ranjan committed:
Updated JavaDocs for GetSourceSetsAction.java

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java:
```diff
@@ -0,0 +1,110 @@
+      var dependencies = new HashSet<BuildTargetDependency>();
+      for (File file : entry.getValue()) {
+        GradleSourceSet otherSourceSet = outputsToSourceSet.get(file);
+        if (otherSourceSet != null) {
```
Is it possible that the otherSourceSet is the current sourceset being iterated?

I'm asking because I'm thinking that if it's necessary to check it here.

Tanish-Ranjan replied:
I don't think a source set can have a dependency on it's own. Like a source set to output an artifact that exists in it's own compile classpath? It shouldn't be possible I believe. However I would let the Gradle community confirm it.

jdneo replied:
At least it do no harm.

Tanish-Ranjan replied:
Okay 👍

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java:
```diff
@@ -0,0 +1,110 @@
+  private void fetchModels(
+      BuildController buildController,
+      GradleBuild build,
+      List<String> traversedProjects,
```
If it's to avoid cyclic dependencies, HashSet seems better.

Tanish-Ranjan replied:
I will wait for other comments to get addressed first. Then I will push a commit resolving all of them.

jdneo commented on plugin/src/main/java/com/microsoft/java/bs/gradle/plugin/SourceSetsModelBuilder.java:
```diff
@@ -108,10 +112,13 @@ public Object buildAll(String modelName, Project rootProject) {
         List<File> compileClasspath = new LinkedList<>(sourceSet.getCompileClasspath().getFiles());
```
I'm thinking that if we need to try-catch here. Though the method signature does not declare that it will throw exception.

Will `sourceSet.getCompileClasspath().getFiles()` throw RuntimeException? (maybe a question to @donat @reinsch82 @hegyibalint)

reinsch82 replied:
I don't see anything that would throw there.

jdneo replied:
Before using build action, this line of code throws an exception, see: https://github.com/microsoft/build-server-for-gradle/pull/122#issuecomment-2031099285

So, I'm thinking that wrapping a try-catch for RuntimeException/GradleException

Tanish-Ranjan replied:
Yes, it was the case then. [FileCollection#getFiles](https://docs.gradle.org/current/javadoc/org/gradle/api/file/FileCollection.html#getFiles()) used to throw an exception.

jdneo replied:
I'll handle it in https://github.com/microsoft/build-server-for-gradle/pull/163

Tanish-Ranjan committed:
Addressing comments

jdneo commented on model/src/main/java/com/microsoft/java/bs/gradle/model/GradleSourceSetsMetadata.java:
```diff
@@ -0,0 +1,15 @@
+ * Provides necessary information to build SourceSets.
+ */
+public interface GradleSourceSetsMetadata extends Serializable {
+  Map<GradleSourceSet, List<File>> getGradleSourceSets();
```
`getGradleSourceSets` -> `getGradleSourceSetToClasspath` ?

jdneo commented on model/src/main/java/com/microsoft/java/bs/gradle/model/GradleSourceSetsMetadata.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Map;
+
+/**
+ * Provides necessary information to build SourceSets.
```
Add more description to explain the purpose of this class?

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java:
```diff
@@ -0,0 +1,110 @@
+      var dependencies = new HashSet<BuildTargetDependency>();
+      for (File file : entry.getValue()) {
+        GradleSourceSet otherSourceSet = outputsToSourceSet.get(file);
+        if (otherSourceSet != null && entry.getKey() != otherSourceSet) {
```
Use `Objects.equals()` instead of `!=`

jdneo commented on plugin/src/main/java/com/microsoft/java/bs/gradle/plugin/SourceSetsModelBuilder.java:
```diff
@@ -43,11 +42,15 @@
   @Override
   public Object buildAll(String modelName, Project rootProject) {
+
+    Map<GradleSourceSet, List<File>> gradleSourceSets = new HashMap<>();
```
rename to `sourceSetToClasspath`?

Tanish-Ranjan committed:
Improved readability and clarity of GradleSourceSetsMetadata
Changes:
- Refactored variable and method names to reflect their purpose clearly.
- Updated JavaDoc comments for better explanation.

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java:
```diff
@@ -0,0 +1,111 @@
+  private void fetchModels(
+      BuildController buildController,
+      GradleBuild build,
+      HashSet<String> traversedProjects,
```
nit: `HashSet<String>` -> `Set<String>`

To align with the below params, since they are Map, not HashMap

Tanish-Ranjan committed:
Updated GetSourceSetsAction#fetchModels parameter type for consistency

jdneo reviewed (approved):
LGTM.

Tested with [Gradle official composite build sample](https://docs.gradle.org/current/samples/sample_composite_builds_basics.html) and [a composite build project](https://github.com/microsoft/vscode-gradle/issues/1435) that provided by a user. Both works as expected.

Thank you @Tanish-Ranjan and @Arthurm1 for the contribution, good job!

Tanish-Ranjan commented:
Big thanks to @Arthurm1 for the initial insights into the tooling API! Your guidance gave this PR a significant head start. Additionally, a huge thanks to all the gsoc mentors (@jdneo @donat @hegyibalint @reinsch82) who provided support throughout this process. Your help was invaluable in getting this completed so quickly.

jdneo commented:
Ok, so after some investigation, I found that the current solution has some problems:

1. See #169 and #170
2. For the sample project provided in https://github.com/microsoft/vscode-gradle/issues/1435, all the sourceset from two subproject will have the same project path `:`, this is actually not correct, we need to use the build tree path to differentiate them. Otherwise, the build target dependency resolution in `BuildTargetManager` might be wrong.
